contributors/design-proposals/resource-class.md

```diff
@@ -0,0 +1,363 @@
+application pod may perform well only with resources which have a certain
+capability, for example GPUs greater than version 'V'. OIR does not allow such,
+metadata based, selection/filtering of resources.
+The current model requires identity mapping between available resources and requested
+resources. Identity mapping means there must be a one-to-one mapping between the
+resource reference used in the spec to request the resource and resource reference
+used to advertise the resource availablity.
+
+Since 'CPU' and 'Memory' are resources that are available across all kubernetes
+deployments and need no metadata to describe the resource type any further,
+the current user facing API (Pod Specification) remains portable as long as pod
+requests only CPU and Memory. However the current model cannot support
+complex resources like GPUs, ASICs, NICs, local storage, etc.
+To support heterogeneity, portability and management at scale, such resources
+must be represented(advertised) in a form which allows metadata inclusion and a
+metadata based resource selection mechanism.
+
+_GPU Integration Example:_
+  * [Enable "kick the tires" support for Nvidia GPUs in COS](https://github.com/kubernetes/kubernetes/pull/45136)
+  * [Extend experimental support to multiple Nvidia GPUs](https://github.com/kubernetes/kubernetes/pull/42116)
+
+_Kubernetes Meeting Notes On This:_
+  * [Meeting notes](https://docs.google.com/document/d/1Qg42Nmv-QwL4RxicsU2qtZgFKOzANf8fGayw8p3lX6U/edit#)
+  * [Better Abstraction for Compute Resources in Kubernetes](https://docs.google.com/document/d/1666PPUs4Lz56TqKygcy6mXkNazde-vwA7q4e5H92sUc)
+  * [Extensible support for hardware devices in Kubernetes (join kubernetes-dev@googlegroups.com for access)](https://docs.google.com/document/d/1LHeTPx_fWA1PdZkHuALPzYxR0AYXUiiXdo3S0g2VSlo/edit)
+
+## Use Cases
+
+  * I want to have a compute resource type, 'Resource Class',
+    which can be created with meaningful and portable names. This compute
+    resource can hold additional metadata as well, for example:
+    * `nvidia.gpu.high.mem` is the name and metadata is memory greater than 'X' GB.
+    * `fast.nic` is the name and associated metadata is bandwidth greater than
+      'B' gbps.
+  * If I request a resource `nvidia.gpu.high.mem` for my pod, any 'nvidia-gpu'
+    type device which has memory greater than or equal to 'X' GB, should be able
+    to satisfy this request, independent of other device capabilities such as
+    'version' or 'nvlink locality' etc.
+  * Similarly, if I request a resource `fast.nic`, any nic device with speed
+    greater than 'B' gbps should be able to meet the request.
+  * I want a rich metadata selection interface where operators like 'Eq' for
+    'equals to', 'Lt' for 'less than', 'LtEq' for 'less than equal to', 'Gt' for
+    'greater than', 'GtEq' for 'greater than and equal to' and 'In' for
+    'a set of accepted values' are supported on the compute resource metadata.
```

Comment:
I'd recommend starting with boolean capabilities and only adding the Lt/Gt rich stuff if the need arises.

Reply:
Lt/Gt is necessary to tackle the example scenario you suggested above:
>nvidia.com/gpu/tesla.k80
nvidia.com/gpu/tesla.k40
nvidia.com/gpu/tesla.c2075
nvidia.com/gpu/tesla.c2070
nvidia.com/gpu/tesla.c2050

For such cases, we can have a fewer resource classes classified using Lt/Gt/IN on the, for example, model number. one resource class `tesla-k-series` : 
```
kind: ResourceClass
metadata:
  name: tesla-k-series
spec:
  resourceSelector:
    - matchExpressions:
        - key: "Kind"
          operator: "In"
          values:
            - "nvidia-gpu"
        - key: "model"
          operator: "In"
          values:
            - "k80", "k40"
```
cluster admin must have knowledge that how device plugin is going to represent properties of device the `Device` api object and correspondingly should create abstracted resource classes, to be offered to users for consuption.

Reply:
The In operator is fine -- it's just an OR operator which works fine when the qualitative strings are considered boolean values -- i.e. the provider of the resource either has the capability or it doesn't have the capability.

The example scenario I put above with nvidia.com/gpu/tesla.k80, etc was something I was recommending NOT to get into :)

I would not recommend using vendor product lines as the things that you match against because:

* The product/brand line from a vendor inevitably gets changed as marketing departments change :)
* The product/vendor/brand isn't necessarily what the *end user* is interested in. The end user wants a GPU (or # of vGPUs) that support X number of display heads and supports some specific software/hardware firmware version (e.g. supports OpenCL 2.0). The user doesn't care which vendor product is called. The user just wants a specific set of capabilities and requested resource amounts.

Reply:
Sure.
That was just an example. It could be any property. Proposal provides approach to use any arbitrary key-value map of device properties. This proposal provides infra and it would depend on device plugin and cluster admin that how best/worse, it is used.